modules/dnn/src/layers/nonzero_layer.cpp

```diff
@@ -0,0 +1,243 @@
+    {
+        return backendId == DNN_BACKEND_OPENCV;
+    }
+
+    bool getMemoryShapes(const std::vector<MatShape>& inputs, const int /*requiredOutputs*/,
+                         std::vector<MatShape>& outputs, std::vector<MatShape>& /*internals*/) const CV_OVERRIDE
+    {
+        CV_Assert(inputs.size() == 1);
+        const MatShape& in = inputs[0];
+        CV_Assert(!in.empty());
+
+        int rank = (int)in.size();
+        MatShape out = shape(rank, -1);
+        outputs.assign(1, out);
+        return false;
+    }
+
+    void getTypes(const std::vector<MatType>& /*inputs*/, const int requiredOutputs,
+                  const int requiredInternals, std::vector<MatType>& outputs,
+                  std::vector<MatType>& internals) const CV_OVERRIDE
+    {
+        outputs.assign(requiredOutputs, CV_64S);
+    }
+
+    void forward(InputArrayOfArrays in_arr, OutputArrayOfArrays out_arr, OutputArrayOfArrays) CV_OVERRIDE
+    {
+        CV_Assert(in_arr.size().area() == 1);
+        const Mat X = in_arr.getMat(0);
+        CV_Assert(X.data != nullptr);
+
+        const int rank = X.dims;
+        std::vector<int> dims(rank), strides(rank);
+        for (int i = 0; i < rank; ++i) dims[i] = X.size[i];
+
+        if (rank > 0) {
+            strides.back() = 1;
+            for (int i = rank - 2; i >= 0; --i)
+                strides[i] = strides[i + 1] * dims[i + 1];
+        }
+
+        size_t total = X.total();
+        const int nstripes = 16;
+
+        const int depth = CV_MAT_DEPTH(X.type());
+        std::vector<size_t> nzcounts;
+        std::vector<size_t> nzstart;
+        size_t nnz = 0;
+        switch (depth)
+        {
+        case CV_8U:
+        case CV_8S:  { nnz = computeNonZeroCountsPerStripe<uchar>(X.ptr<uchar>(), total, nstripes, nzcounts); break; }
+        case CV_16U:
+        case CV_16S: { nnz = computeNonZeroCountsPerStripe<uint16_t>(X.ptr<uint16_t>(), total, nstripes, nzcounts); break; }
+        case CV_32U:
+        case CV_32S: { nnz = computeNonZeroCountsPerStripe<uint32_t>(X.ptr<uint32_t>(), total, nstripes, nzcounts); break; }
+        case CV_64U:
+        case CV_64S: { nnz = computeNonZeroCountsPerStripe<uint64_t>(X.ptr<uint64_t>(), total, nstripes, nzcounts); break; }
+        case CV_32F: { nnz = computeNonZeroCountsPerStripe<float>(X.ptr<float>(), total, nstripes, nzcounts); break; }
+        case CV_64F: { nnz = computeNonZeroCountsPerStripe<double>(X.ptr<double>(), total, nstripes, nzcounts); break; }
+        case CV_16F: { nnz = computeNonZeroCountsPerStripe<uint16_t>(X.ptr<uint16_t>(), total, nstripes, nzcounts, [](uint16_t v){ return isNonZeroF16(v); }); break; }
+        case CV_16BF:{ nnz = computeNonZeroCountsPerStripe<uint16_t>(X.ptr<uint16_t>(), total, nstripes, nzcounts, [](uint16_t v){ return isNonZeroF16(v); }); break; }
+        case CV_Bool:{ nnz = computeNonZeroCountsPerStripe<uchar>(X.ptr<uchar>(), total, nstripes, nzcounts); break; }
```

Comment:
CV_Bool branch looks identical to CV_8U/CV_8S;
CV_16F and CV_16BF are identical as well, please, reduce the code further